src/Main.elm

```diff
@@ -4,76 +4,393 @@ import Browser
+import Shared.Button as Button
+import Shared.Editable as Editable
+import Shared.Icon as Icon
+import Shared.Value as Value
 
 
-
----- MODEL ----
+type alias Tag =
+    { name : Editable.Editable String
+    , value : Editable.Editable String
+    }
 
 
 type alias Model =
-    {}
+    { tags : List Tag
+    , newTagForm : Tag
+    , globalError : Maybe String
+    }
 
 
 init : ( Model, Cmd Msg )
 init =
-    ( {}, Cmd.none )
+    let
+        ( tags, error ) =
+            decodeTagsWithMaybeError
+    in
+    ( { tags = tags, newTagForm = emptyNewTagForm, globalError = error }, Cmd.none )
+
+
+emptyNewTagForm : Tag
+emptyNewTagForm =
+    { name = Value.toValue "" |> Editable.valueToEditable
+    , value = Value.toValue "" |> Editable.valueToEditable
+    }
+
+
+tagDecoder : Decode.Decoder Tag
+tagDecoder =
+    Decode.succeed Tag
+        |> Pipeline.required "name" (Decode.map Editable.toReadonly Decode.string)
+        |> Pipeline.optional "value" (Decode.map Editable.toReadonly Decode.string) (Editable.toReadonly "")
 
 
+decodeTagsWithMaybeError : ( List Tag, Maybe String )
+decodeTagsWithMaybeError =
+    let
+        decodeTags =
+            Decode.field "tags" (Decode.list tagDecoder)
+    in
+    case Decode.decodeString decodeTags Data.userGroup of
+        Ok decodedTags ->
+            ( decodedTags, Nothing )
 
----- UPDATE ----
+        Err decodeError ->
+            ( [], Just ("Error decoding data: " ++ Decode.errorToString decodeError) )
 
 
 type Msg
-    = NoOp
+    = UpdateNewName String
+    | UpdateNewValue String
+    | ClickedAddTag
+    | ClickedRemoveTag String
+    | InsertedTagValue String String
+    | ClickedEditTagValue Bool String
+    | ClickedReloadData
+    | NoOp
 
 
 update : Msg -> Model -> ( Model, Cmd Msg )
-update _ model =
-    ( model, Cmd.none )
+update msg model =
+    case msg of
+        UpdateNewName nameString ->
+            let
+                updateName tag =
+                    { tag | name = Editable.updateIfEditable tag.name nameString }
+            in
+            ( { model | newTagForm = updateName model.newTagForm }, Cmd.none )
+
+        UpdateNewValue valueString ->
+            let
+                updateValue tag =
+                    { tag | value = Editable.updateIfEditable tag.value valueString }
+            in
+            ( { model | newTagForm = updateValue model.newTagForm }, Cmd.none )
+
+        ClickedAddTag ->
+            let
+                validateNameValue nameValue =
+                    let
+                        nameString =
+                            Value.fromValue nameValue
+                    in
+                    if String.isEmpty nameString then
+                        Value.addError "Name is required" nameValue
+
+                    else if List.any (.name >> Editable.valueFromEditable >> Value.fromValue >> (==) nameString) model.tags then
+                        Value.addError "Duplicate name" nameValue
```

Comment:
The nesting of `let .. in` inside `let .. in` is a codesmell to me at it might be quite challenging to follow which value is from where and it adds complexity to overall code readability. Would it be possible to refactor it out? How?